sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/Schema.java

```diff
@@ -953,6 +1012,338 @@ public int hashCode() {
+          return false;
+        }
+        Option option = (Option) o;
+        return Objects.equals(type, option.type) && Objects.equals(value, option.value);
+      }
+
+      @Override
+      public int hashCode() {
+        return Objects.hash(type, value);
+      }
+    }
+
+    public static class Builder {
+      private Map<String, Option> options;
+
+      Builder(Map<String, Option> init) {
+        this.options = new HashMap<>(init);
+      }
+
+      Builder() {
+        this(new HashMap<>());
+      }
+
+      public Builder setByteOption(String optionName, Byte value) {
+        setOption(optionName, FieldType.BYTE, value);
+        return this;
+      }
+
+      public Builder setBytesOption(String optionName, byte[] value) {
+        setOption(optionName, FieldType.BYTES, value);
+        return this;
+      }
+
+      public Builder setInt16Option(String optionName, Short value) {
+        setOption(optionName, FieldType.INT16, value);
+        return this;
+      }
+
+      public Builder setInt32Option(String optionName, Integer value) {
+        setOption(optionName, FieldType.INT32, value);
+        return this;
+      }
+
+      public Builder setInt64Option(String optionName, Long value) {
+        setOption(optionName, FieldType.INT64, value);
+        return this;
+      }
+
+      public Builder setDecimalOption(String optionName, BigDecimal value) {
+        setOption(optionName, FieldType.DECIMAL, value);
+        return this;
+      }
+
+      public Builder setFloatOption(String optionName, Float value) {
+        setOption(optionName, FieldType.FLOAT, value);
+        return this;
+      }
+
+      public Builder setDoubleOption(String optionName, Double value) {
+        setOption(optionName, FieldType.DOUBLE, value);
+        return this;
+      }
+
+      public Builder setStringOption(String optionName, String value) {
+        setOption(optionName, FieldType.STRING, value);
+        return this;
+      }
+
+      public Builder setDateTimeOption(String optionName, ReadableDateTime value) {
+        setOption(optionName, FieldType.DATETIME, value);
+        return this;
+      }
+
+      public Builder setBooleanOption(String optionName, Boolean value) {
+        setOption(optionName, FieldType.BOOLEAN, value);
+        return this;
+      }
+
+      public <V> Builder setArrayOption(String optionName, FieldType arrayType, List<V> value) {
+        setOption(optionName, arrayType, value);
+        return this;
+      }
+
+      public <K, V> Builder setMapOption(String optionName, FieldType mapType, Map<K, V> value) {
+        setOption(optionName, mapType, value);
+        return this;
+      }
+
+      public Builder setRowOption(String optionName, Row value) {
+        setOption(optionName, FieldType.row(value.getSchema()), value);
+        return this;
+      }
+
+      public Builder setOption(String optionName, FieldType fieldType, Object value) {
+        if (value == null) {
```

Comment:
why special case null?

Reply:
Options should never be null, if a null is set it's equivalent to removing the option. Or do you prefer and Exception?